xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1354 @@
+		t.Fatalf("timeout while waiting for an NewSubConn call on the ClientConn")
+	case <-tcc.NewSubConnCh:
+	}
+
+	sc2, err := gsb.balancerPending.Balancer.(*mockBalancer2).newSubConn([]resolver.Address{}, balancer.NewSubConnOptions{}) // Will eventually get back a SubConn with an identifying property id 2
+	if err != nil {
+		t.Fatalf("error constructing newSubConn in gsb: %v", err)
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an NewSubConn call on the ClientConn")
+	case <-tcc.NewSubConnCh:
+	}
+
+	currBal := gsb.balancerCurrent.Balancer.(*mockBalancer1)
+	pendBal := gsb.balancerPending.Balancer.(*mockBalancer2)
+
+	// Closing the Graceful Switch Load Balancer should lead to removing any
+	// created SubConns, and closing both the current and pending Load Balancer.
+	gsb.Close()
+
+	// The order of SubConns the Graceful Switch Balancer tells the Client Conn
+	// to remove is non deterministic, as it is stored in a map. However, the
+	// first SubConn removed should be either sc1 or sc2.
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateAddresses call on the ClientConn")
+	case sc := <-tcc.RemoveSubConnCh:
+		if !cmp.Equal(sc1, sc, cmp.AllowUnexported(testutils.TestSubConn{})) {
+			if !cmp.Equal(sc2, sc, cmp.AllowUnexported(testutils.TestSubConn{})) {
+				t.Fatalf("RemoveSubConn, want either %v or %v, got %v", sc1, sc2, sc)
+			}
+		}
+	}
+
+	// The Graceful Switch Balancer should then tell the Client Conn to remove
+	// the other SubConn.
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateAddresses call on the ClientConn")
+	case sc := <-tcc.RemoveSubConnCh:
+		if !cmp.Equal(sc1, sc, cmp.AllowUnexported(testutils.TestSubConn{})) {
+			if !cmp.Equal(sc2, sc, cmp.AllowUnexported(testutils.TestSubConn{})) {
+				t.Fatalf("RemoveSubConn, want either %v or %v, got %v", sc1, sc2, sc)
+			}
+		}
+	}
+
+	// The current balancer should get closed as a result of the graceful switch balancer being closed.
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := currBal.waitForClose(ctx); err != nil {
+		t.Fatalf("error in close: %v", err)
+	}
+	// The pending balancer should also get closed as a result of the graceful switch balancer being closed.
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := pendBal.waitForClose(ctx); err != nil {
+		t.Fatalf("error in close: %v", err)
+	}
+
+	// Also, once this event happens, trying to do anything else on both codepaths
+	// should be a no-op. (Should I try and have an event happen from a lower level
+	// balancer? Is this possible?)
```

Comment:
What do you mean "a lower level balancer"?

Reply:
Thinking about this, I have this covered in other parts of this file (i.e. a closed balancers updates should be eaten and not forwarded up to the ClientConn).